astropy/units/decorators.py

```diff
@@ -115,6 +117,28 @@ def wrapper(*func_args, **func_kwargs):
                 # If the target unit is empty, then no unit was specified so we
                 # move past it
                 if target_unit is not funcsigs.Parameter.empty:
+
+                    if isinstance(target_unit, str):
+
+                        try: # unit passed in as a string
+                            target_unit = Unit(target_unit)
+                            str_target_unit = target_unit.to_string()
+                        except ValueError:
+                            # user specified a physical type instead of a unit
+                            try:
+                                physical_type_id = _unit_physical_mapping[target_unit]
+                                str_target_unit = target_unit
+                            except KeyError:
+                                raise ValueError("Invalid physical type '{0}'."
```

Comment:
The error message now presumes the string was meant to be a physical type; just change to "Invalid unit of physical type '{0'}" will do, I think.